File: laberinto_final.py
import pytest

def eliminar_enter(laberinto):
    '''Función que recibe el laberinto y elimina el enter de cada reglon
    eliminar_enter: Lista de string -> None
    Recibe una lista de strings que representan un laberinto y elimina el '\n' de cada string.'''
    
    ANCHO = len(laberinto[0])
    ALTO = len(laberinto)
    ancho = len(laberinto[0])
    
    for i in range(0, len(laberinto)):
        laberinto[i] = laberinto[i][:(ancho-1)]
    return laberinto

def lector_coordenadas():
    '''Función que lee el archivo de coordenadas
    lector_coordenadas: None -> Lista de Strings
    La función no recibe nada. Abre el achivo donde se encuentran las coordenadas y
    devuelve el laberinto representado por una lista de strings. '''
    
    archivo = open('laberinto.txt','r')   #abre el archivo
    laberinto = archivo.readlines()
    archivo.close()
    
    laberinto = eliminar_enter(laberinto) #elimina el '\n' de cada fila
    
    return laberinto



def mover(pos_actual, direccion):
    '''Función que se mueve según la dirección
    mover: Tupla, String -> Tupla
    Recibe una tupla que representa la posición actual y un string que representa
    la dirección a donde se quiere ir. Devuelve una tupla que representa la nueva posición.'''
    
    if direccion == 'derecha':
        return (pos_actual[0],pos_actual[1]+1)
    elif direccion == 'izquierda':
        return (pos_actual[0], pos_actual[1]-1)
    elif direccion == 'arriba':
        return (pos_actual[0]-1, pos_actual[1])
    else:                                       #se mueve hacia abajo
        return (pos_actual[0]+1, pos_actual[1])

def mover_test():
    '''Función de prueba de mover'''
    assert mover((5,5),'derecha') == (5,6)
    assert mover((5,5),'izquierda') == (5,4)
    assert mover((5,5),'arriba') == (4,5)
    assert mover((5,5), '') == (6,5)
        

def puedo_mover(pos_actual,laberinto, ANCHO, ALTO):
    '''Función que nos dice cuantas veces nos podemos mover.
    puedo_mover: Tupla, Lista de Strings, Entero, Entero -> Entero
    Recibe una tupla que representa la posición actual, una lista de strings
    que representa a el laberinto, y dos enteros que representan cuantas columnas y filas hay en
    el laberinto. Devuelve un entero que representa cuantos posbiles movimientos hay desde
    la posición actual.'''

    i = 0 #contador de posibles movimientos

    derecha = mover(pos_actual, 'derecha')
    if  derecha[1] < ANCHO and laberinto[derecha[0]][derecha[1]] != '1' and laberinto[derecha[0]][derecha[1]] !='3':
        i = i+1
        

    izquierda = mover(pos_actual, 'izquierda')
    if  izquierda[1] >= 0 and laberinto[izquierda[0]][izquierda[1]] != '1' and laberinto[izquierda[0]][izquierda[1]] !='3':
        i = i+1

    arriba = mover(pos_actual, 'arriba')
    if arriba[0] >= 0  and arriba[1] <= ANCHO-1 and laberinto[arriba[0]][arriba[1]] != '1' and laberinto[arriba[0]][arriba[1]] != '3':
        i = i+1

    abajo = mover(pos_actual, 'abajo')
    if abajo[0] < ALTO and laberinto[abajo[0]][abajo[1]] != '1' and laberinto[abajo[0]][abajo[1]] != '3':
        i = i+1
        
    return i

def puedo_mover_test():
    '''Función de prueba de la funcion puedo_mover'''
    
    lab = [['030'],
           ['100'],
           ['000']]
    assert puedo_mover((0,0), lab, 3, 3) == 0
    assert puedo_mover((1,1), lab, 3, 3) == 2
    assert puedo_mover((2,1), lab, 3, 3) == 3
        


def avanzar(pos_actual, laberinto, ANCHO, ALTO):
    '''Función que hace que se mueva la posición actual.
    avanzar: Tupla, Lista de Strings, Entero, Entero -> Tupla
    Recibe una tupla que representa la posición actual, una lista de strings que representa
    al laberinto y dos enteros que representan cuantas columnas y filas hay en el laberinto.
    Devuelve una tupla que representa la posición a la que se puede avanzar.'''

    derecha = mover(pos_actual, 'derecha')
    izquierda = mover(pos_actual, 'izquierda')
    arriba = mover(pos_actual, 'arriba')
    abajo = mover(pos_actual, 'abajo')
    
    if  derecha[1] < ANCHO and laberinto[derecha[0]][derecha[1]] != '1' and laberinto[derecha[0]][derecha[1]] !='3':
        return derecha
    
    elif izquierda[1] >= 0 and laberinto[izquierda[0]][izquierda[1]] != '1' and laberinto[izquierda[0]][izquierda[1]] !='3':
        return izquierda
    
    elif arriba[0] >= 0  and arriba[1] <= ANCHO-1 and laberinto[arriba[0]][arriba[1]] != '1' and laberinto[arriba[0]][arriba[1]] != '3':
        return arriba
    
    else:    #si llego hasta acá es porque se puede mover por lo menos una vez
        return abajo
    

def avanzar_test():
    '''Función de prueba de la función avanzar'''

    laberinto = [['0010'],
                 ['0110'],
                 ['0101'],
                 ['0000']]
    
    assert avanzar((0,0), laberinto, 4, 4) == (0,1) #derecha
    assert avanzar((1,0), laberinto, 4, 4) == (0,0) #izquierda
    assert avanzar((1,3), laberinto, 4, 4) == (0,3) #arriba
    assert avanzar((2,2), laberinto, 4, 4) == (3,2) #abajo
    
    

def reemplazar_numeros(fila, posicion, ANCHO, caracter):
    '''Función que modifica una fila del laberinto en determinada posición.
    reemplazar_numeros: String, Entero, Entero, String -> String 
    Recibe un string que representa la fila del laberinto, un entero que representa la posición de la fila que
    queremos modificar, un entero que representa el largo 
    queremos poner a el número en la fila'''


    if posicion == ANCHO-1:            #si la posición está al final de la fila
        return fila[:posicion]+caracter
    elif posicion == 0:                #si la posición está al principio de la fila
        return caracter+fila[1:]
    else:
        return fila[:posicion]+caracter+fila[posicion+1:]
    
def resolver_laberinto(pos_actual, laberinto, camino, ANCHO, ALTO):
    '''Función que resuelve el laberinto.
    resolver_laberinto: Tupla, Lista de Strings, Lista de tuplas, Entero, Entero -> Lista de tuplas
    Recibe una tupla que representa la posición actual, una lista de strings que representa al laberinto,
    una lista de tuplas que representa el camino que fue haciendo y dos enteros que representan el tamaño del laberinto.'''

    longitud_camino = len(camino)
    
    if laberinto[pos_actual[0]][pos_actual[1]] == '2': 
        camino.append(pos_actual)
        solucion=camino[:]
        return solucion
    
    elif puedo_mover(pos_actual,laberinto, ANCHO, ALTO) == 0:  #si no me puedo mover
        
        if longitud_camino > 0: 
            laberinto[pos_actual[0]] = reemplazar_numeros(laberinto[pos_actual[0]], pos_actual[1], ANCHO, '1')
            return resolver_laberinto(camino[longitud_camino-1], laberinto, camino[:longitud_camino-1], ANCHO, ALTO)
        else:
            return []
    else:
        laberinto[pos_actual[0]] = reemplazar_numeros(laberinto[pos_actual[0]], pos_actual[1], ANCHO, '3')  #pone un 3 para saber que ya pasó por allí
        camino.append(pos_actual)
        return resolver_laberinto(avanzar(pos_actual,laberinto, ANCHO, ALTO),laberinto, camino, ANCHO, ALTO)

def resolver_laberinto_test():
    '''Función de prueba de la función resolver_laberinto'''

    assert resolver_laberinto((0,1), [['0210'],['0100']], [(0,0)], 4, 2) == [(0,0),(0,1)]
    assert resolver_laberinto((0,0), [['01'],['10']], [], 2,2) == []
    assert resolver_laberinto((0,0), [['0010'],['1001'],['1101'],['0121']], [], 4,4) == [(0,0), (0,1),(1,1),(2,1),(3,1)]

    
def escritor_solucion(solucion, ANCHO, ALTO):
   ''' Función que crea un archivo con la solucion.
   escritor_solucion: Lista de tuplas, Entero, Entero -> None
   Recibe una lista de tuplas que representa la solución del laberinto, y dos enteros que representan
   el tamaño de las columnas y de las filas del laberinto.
   Si la solución era una lista vacía crea un archivo con un espacio ('\n'). Si no, crea un archivo
   con cada uno de los pasos hasta el objetivo incluido. '''
   
   archivo = open('solucion.txt', 'w')
   if (solucion == []):  
      archivo.write(' ')
      print('El archivo no tiene solución')
   else:
       for i in range(len(solucion)):
            archivo.write(str(solucion[i][0]) + ' ' + str(solucion[i][1]) + '\n')
   archivo.close()
   print('El archivo se ha creado con éxito.')
   


def principal():
                                
    '''Función principal.
    principal: None -> None
    Lee un archivo con el laberinto y crea un archivo con los pasos que hay que seguir
    desde la posición de comienzo (0,0) hasta el objetivo.'''

    laberinto = lector_coordenadas()
    
    camino = []
    pos_actual = (0,0)
    ALTO = len(laberinto)
    ANCHO = len(laberinto[0])

    if laberinto[pos_actual[0]][pos_actual[1]] == '1' or puedo_mover(pos_actual,laberinto, ANCHO, ALTO) == 0: #Si hay una pared o no me puedo mover al principio.
        solucion = []
    else:
        solucion=resolver_laberinto(pos_actual, laberinto, camino, ANCHO, ALTO)
        
    escritor_solucion(solucion, ANCHO, ALTO) #crea un archivo con la solución


principal()

File: laberinto_final.c
#include <stdio.h>
#include <stdlib.h>

// Fución que lee un archivo.
// crea_arreglo: int[] -> int
// Recibe un arreglo de enteros que representan coordenadas, lee un archivo y pasa los datos al arreglo.
// Devuelve un entero que representa cuantos elementos hay en el arreglo.
int crea_coordenadas(int coordenadas[44]) {

  FILE*archivo = fopen("coordenadas.txt", "r"); //abre el archivo en modo solo lectura

  int i=0; //cuenta cuantos elementos hay en el arreglo.
  while (!feof(archivo) && i < 44) {
    fscanf(archivo,"%d", &coordenadas[i]);
    i++;
  }

  fclose(archivo);

  return i-1; //Menos uno ya que el arreglo me lo devuelve con un 0 final.
}

// Función que se fija que las coordenadas estén dentro del laberinto.
// se_pueden_coordenadas: int[],int, int -> int
// Recibe un arreglo de números que representa las coordenadas y dos enteros que representan
//el alto y el ancho del laberinto.
// Dada las coordenadas, las filas y las columnas, se fija que todas las coordenadas
//estén dentro del laberinto, devuelve 0 si se puede y 1 si no.
int se_pueden_coordenadas(int coordenadas[44], int cantidad_coordenadas, int ALTO, int ANCHO) {

  int i;
  int bandera = 0;
    for(i=2; i<cantidad_coordenadas && !bandera; i++) {
      if (i%2 == 0){
        if (coordenadas[i] >= ALTO) {
          bandera =1;
        }
      }
      else{
        if (coordenadas[i] >= ANCHO) {
          bandera = 1;
        }
      }
    }
  return bandera;
  }

// Función que se fija si hay números negativos en el arreglo.
// hay_negativos: int[], int -> int
// Recibe un arreglo de números que representan coordenadas y un número que representa la cantidad
//de elementos que hay en el arreglo. Dada las coordenadas, se fija si hay algún número negativo,
//devolviendo 1 en ese caso. Si no, devuelve un 0.
int hay_negativos(int coordenadas[44], int cantidad_coordenadas) {

  int i;
  int bandera = 0;
  for(i=0; i < cantidad_coordenadas && !bandera  ; i++) {
    if (coordenadas[i] < 0) {
      bandera = 1;
    }
  }
  return bandera;
}


// Función que coloca las casillas vacias en el laberinto.
// colocar_vacios: int[][], int, int -> None
// Recibe un arreglo de enteros que representa el laberinto y entereos que representan
//el ancho y el largo del laberinto. En base a eso va a llenar todo de casillas vacías.
void colocar_vacios(int laberinto[][15], int ALTO, int ANCHO) {

  int i,j;
  int contador = 1;
  for(i=0; i<ALTO; i++) {
    for(j=0; j<ANCHO; j++) {
      laberinto[i][j] = 0;
    }
  }
}



// Función que coloca las paredes en el laberinto.
// colocar_paredes: int[], int[][], int -> None
// Función que recibe un arreglo de enteros que representa coordenadas, un arreglo de arreglo de enteros
//que representa el laberinto y un número entero que representa la cantidad de coordenadas.
void colocar_paredes(int coordenadas[44], int laberinto[][15], int cantidad_coordenadas) {

  int i=4 ;//índices de las coordenadas
  while(i<cantidad_coordenadas) {
    if (laberinto[coordenadas[i]][coordenadas[i+1]] != 2){ //si no está el objetivo
        laberinto[coordenadas[i]][coordenadas[i+1]] = 1;  //se coloca la pared
        i=i+2;
    }
    else{
      i=i+2;
    }
  }
}

void crea_laberinto(int laberinto[][15], int ALTO, int ANCHO){
  //Función que crea el archivo con el laberinto
  // crea_laberinto: int[][], int, int -> None
  // Recibe un arreglo de arreglo de enteros que representa el laberinto y dos entereos
  //que representan el alto y el ancho del laberinto, y crea un archivo con el laberinto.

	FILE* archivo = fopen("laberinto.txt", "w"); //w:solo escritura

  int f=0; //filas
  while (f < ALTO) {
    int c=0; //columnas
    while (c < ANCHO){
      while(c < ANCHO-1){
        fprintf(archivo,"%d", laberinto[f][c]);
        c++;
      }
      fprintf(archivo,"%d\n", laberinto[f][c]);
      c++;
    }
    f++;
  }
  fclose (archivo);
}



//Función principal.
//Lee un archivo de coordenadas y devuelve un archivo (si se puede) con números que
//representan un laberinto. No se va a poder crear el archivo de texto cuando haya números
//negativos en las coordenadas, cuando se quieran poner paredes en coordenadas mayores al tamaño
//del laberinto o cuando no haya verdaderamente un grupo de coordenadas.
int main(){

  int coordenadas[44];
  int laberinto[15][15];
  int cantidad_coordenadas = crea_coordenadas(coordenadas);

  int ANCHO = coordenadas[1];
  int ALTO = coordenadas[0];
  int objetivo_f = coordenadas[2];
  int objetivo_c = coordenadas[3];

  if (hay_negativos(coordenadas, cantidad_coordenadas)) {
    printf("El archivo posee números negativos.\n");
  }
  else{
    if (se_pueden_coordenadas(coordenadas,cantidad_coordenadas, ALTO, ANCHO)) {
      printf("Las coordenadas del objetivo no son posibles.\n");
    }
    else{
      if  (cantidad_coordenadas%2 != 0) {
        printf("El archivo no posee coordenadas completas.\n");
      }
      else {
        colocar_vacios(laberinto, ALTO, ANCHO);
        laberinto[objetivo_f][objetivo_c] = 2;  //se coloca el objetivo en el laberinto.
        colocar_paredes(coordenadas, laberinto, cantidad_coordenadas);
        crea_laberinto(laberinto, ALTO, ANCHO);
        printf("Se creó el laberinto exitosamente.\n");
      }
    }
  }
  return 0;
}
